### ISSUE TITLE
Parcel 2: @parcel/transformer-html: Source indices must be monotonic

### ISSUE THREAD

**Issue by colorgold** [parcel-bundler/parcel#6672]
# 🐛 bug report

I upgraded to Parcel 2 but everytime I try to run `npm run dev` like I used to on the old version, I get an error stating `@parcel/transformer-html: Source indices must be monotonic`
## 🎛 Configuration (.babelrc, package.json, cli command)

```js
package.json
{
  "name": "chatmurcianysweb",
  "version": "1.0.0",
  "main": "public/index.js",
  "scripts": {
    "dev": "parcel serve public/index.html",
    "build": "parcel build public/index.html"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@giphy/js-components": "^3.0.2",
    "bindings": "^1.5.0",
    "firebase-functions": "^3.7.0",
    "node-internet-radio": "^0.2.0"
  },
  "devDependencies": {
    "parcel": "^2.0.0-beta.3.1"
  },
  "keywords": [],
  "description": ""
}
```

## 🤔 Expected Behavior

I expect the build to run successfully
## 😯 Current Behavior

I get the following error when I run `npm run dev`:
```
> chatmurcianysweb@1.0.0 dev
> parcel serve public/index.html

Server running at http://localhost:1234
× Build failed.

@parcel/transformer-html: Source indices must be monotonic

  Error: Source indices must be monotonic
  at i.getPosition (D:\WebProjects\ChatMurcianysWeb\node_modules\posthtml-parser\dist\chunk.T6SARJYH.js:1:209)       
  at Object.y [as onclosetag] (D:\WebProjects\ChatMurcianysWeb\node_modules\posthtml-parser\dist\index.js:1:1252)    
  at Parser.closeCurrentTag
  (D:\WebProjects\ChatMurcianysWeb\node_modules\posthtml-parser\node_modules\htmlparser2\lib\Parser.js:237:87)       
  at Parser.onclosetag
  (D:\WebProjects\ChatMurcianysWeb\node_modules\posthtml-parser\node_modules\htmlparser2\lib\Parser.js:210:22)       
  at Tokenizer.emitToken
  (D:\WebProjects\ChatMurcianysWeb\node_modules\posthtml-parser\node_modules\htmlparser2\lib\Tokenizer.js:896:23)    
  at Tokenizer.stateInClosingTagName
  (D:\WebProjects\ChatMurcianysWeb\node_modules\posthtml-parser\node_modules\htmlparser2\lib\Tokenizer.js:253:18)    
  at Tokenizer.parse
  (D:\WebProjects\ChatMurcianysWeb\node_modules\posthtml-parser\node_modules\htmlparser2\lib\Tokenizer.js:652:22)    
  at Tokenizer.write
  (D:\WebProjects\ChatMurcianysWeb\node_modules\posthtml-parser\node_modules\htmlparser2\lib\Tokenizer.js:123:14)    
  at Parser.write
  (D:\WebProjects\ChatMurcianysWeb\node_modules\posthtml-parser\node_modules\htmlparser2\lib\Parser.js:339:24)       
  at w (D:\WebProjects\ChatMurcianysWeb\node_modules\posthtml-parser\dist\index.js:1:1828)
```

## 💁 Possible Solution
I have no idea what the error means

## 🔦 Context
I am unable to update my website because of this issue

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           |2.0.0-beta.3.1
| Node             |v14.17.4
| npm/Yarn         |7.20.3
| Operating System |Windows 10 Home 20H2

**Comment by devongovett** [comment]
Looks like an error in posthtml-parser. Do you have a small code sample that reproduces the issue?

**Comment by colorgold** [comment]
> Looks like an error in posthtml-parser. Do you have a small code sample that reproduces the issue?

Thank you for your reply.

I created a new simple html and javascript project, added Parcel 2 to it, and built the project using `npm run dev`. Everything worked there as expected so I went back to the problematic project and noticed that on my `package.json` I had this:
`"main": "public/index.js",`
while on the new project I have this:
`"main": "public/index.html",`
So, I went ahead and changed the first one to match the second one, but now I get this error:
```
3 |   "version": "1.0.0",
  > 4 |   "main": "public/index.html",
  >   |           ^^^^^^^^^^^^^^^^^^^ File extension must be .js, .mjs, or .cjs
    5 |   "scripts": {
    6 |     "dev": "parcel serve public/index.html",

  ℹ The "main" field is meant for libraries. If you meant to output a .html file, either remove the "main" field or choose a
     different target name.
```

I've tried uninstalling and installing Parcel but nothing seems to work.

**Comment by devongovett** [comment]
The main field cannot point to an HTML file. Did you try removing it as suggested in the error message?

**Comment by colorgold** [comment]
> The main field cannot point to an HTML file. Did you try removing it as suggested in the error message?

I did, and it automatically chooses index.js which when I build it throws the error: `@parcel/transformer-html: Source indices must be monotonic`

**Comment by thewilkybarkid** [comment]
I've just run into the same issue: my source HTML file has a `<hr>`, and it works by making it self-closing (i.e. `<hr/>`). I will see if I can debug/create a sample.

**Comment by thewilkybarkid** [comment]
Managed to reproduce it with just:

```html
<!DOCTYPE html>

<title>Title</title>

<p>Paragraph 1

<hr>

<p>Paragraph 2
```

Looking at posthtml-parser and htmlparser2 to see if the problem's there.

**Comment by thewilkybarkid** [comment]
Narrowed it down to the `sourceLocations` option in posthtml-parser. Seems like htmlparser2 provides an incorrect `endIndex` for the `<hr>` (less than the `startIndex`). Will open an issue/PR there (edit: https://github.com/fb55/htmlparser2/issues/896).

**Comment by iamthelabhras** [comment]
I've run into, and successfully resolved, this same error on three separate occasions.

On the first occasion, the issue was an unclosed `<div>` tag.

On the second occassion, the issue was a duplicate closing </a> tag (e.g. `<a>some link</a>some more text</a>`)

On the third occasion,  the culprit was a pair of opening and closing `<p>` tags bracketing an pair of opening & closing `<details>` tags.

I don't know enough about Parcel's internal workings to venture a guess as to whether the "source indices must be monotonic" is **_always_** a sign of unsemantic / poorly structured HTML, but, every time I've encountered this issue, finding & correcting the faulty underlying HTML has successfully resolved the issue.

Unfortuantely, locating the offending HTML can be an incredibly tedious undertaking since the error Parcel throws doesn't point you in the direction of the offending HTML file, let alone the offending _line_ with the offending HTML file. 

I recommend commenting out or otherwise removing all the links from your project's `index.html`, then adding the links back in one-at-a-time until running Parcel throws a "must be monotonic" error; then, I'd recommend commented out half of your code in the offending file, running Parcel again, and seeing if it throws the same error.  Keep bifurcating until you zero in on the problem.  

It's a tedious strategy for debugging but it's what I did and I was able to resolve this issue without going through my HTML files line by line.  Good luck!

**Comment by colorgold** [comment]
> I've run into, and successfully resolved, this same error on three separate occasions.
> 
> On the first occasion, the issue was an unclosed `<div>` tag.
> 
> On the second occassion, the issue was a duplicate closing tag (e.g. `<a>some link</a>some more text</a>`)
> 
> On the third occasion, the culprit was a pair of opening and closing `<p>` tags bracketing an pair of opening & closing `<details>` tags.
> 
> I don't know enough about Parcel's internal workings to venture a guess as to whether the "source indices must be monotonic" is **_always_** a sign of unsemantic / poorly structured HTML, but, every time I've encountered this issue, finding & correcting the faculty underling HTML has successfully resolved the issue.
> 
> Unfortuantely, locating the offending HTML can be an incredibly tedious undertaking since the error Parcel throws doesn't point you in the direction of the offending HTML file, let alone the offending _line_ with the offending HTML file.
> 
> I recommend commenting out or otherwise removing all the links from your project's `index.html`, then adding the links back in one-at-a-time until running Parcel throws a "must be monotonic" error; then, I'd recommend commented out half of your code in the offending file, running Parcel again, and seeing if it throws the same error. Keep bifurcating until you zero in on the problem.
> 
> It's a tedious strategy for debugging but it's what I did and I was able to resolve this issue without going through my HTML files line by line. Good luck!

Thank you for your advice. I will be doing as you say. Luckily, index.html is the only one I am having a problem with and not the other html files.

**Comment by jbelien** [comment]
I ran into the same issue.
In my case there was no "real" formatting issue, I just had a `<ul></ul>` inside a `<p></p>`. I change the `<p></p>` into a `<div></div>` and it "fixed" it.

**Comment by colorgold** [comment]
> I ran into the same issue.
> In my case there was no "real" formatting issue, I just had a `<ul></ul>` inside a `<p></p>`. I change the `<p></p>` into a `<div></div>` and it "fixed" it.

I also had the same issue as well as having `</br>` instead of `<br />`. I will close the issue since it's all about the HTML code.

**Comment by thewilkybarkid** [comment]
Just to clarify: the bug in htmlparser2 (which can happen on both invalid and _valid_ HTML) means the PostHTML `sourceLocations ` option breaks. Use of this option was added in https://github.com/parcel-bundler/parcel/pull/6247, and as the top feature in https://v2.parceljs.org/blog/rc0/ clearly isn't something that's going to be temporarily reverted.

I've opened https://github.com/posthtml/posthtml-parser/pull/80, but I've had to leave an open comment on htmlparser2 (https://github.com/fb55/htmlparser2/issues/896#issuecomment-894692972) as I'm not sure what the correct fix should be.

As there's no change to Parcel (until dependencies can be updated), I'm just wondering about workarounds: would parsing/rendering the HTML prior to the current PostHTML usage be enough? (i.e. add in all the implied tags)
